### DISCUSSION TITLE
Creation of individual output files for multiapp reset times

### DISCUSSION THREAD

**Discussion by abarun22** [idaholab/moose#20461] [Q&A General] [answered]
Hi Guillaume,
You might've remember this work (https://github.com/idaholab/moose/discussions/18549) which happened quite a while ago. I observed an odd thing concerning the creation of exodus outputs while working on my recent activity on QuadraturePointMultiapp. The specification of list of reset_time's given in the input enables the creation of output files for all these times which would multiply if we are using a large mesh. Is there a possibility of suppressing the creation of such files either by setting options in the input file or by other means?

```
[MultiApps]
  [sub]
    type = QuadraturePointMultiapp
    input_files = 'sub_gs_uniaxial_noCPFE.i'
    output_in_position = true
    execute_on = timestep_end
    reset_time = '0.02 0.03 0.04 0.05'
  []
[]

```
![image](https://user-images.githubusercontent.com/40111479/156363051-14c385f2-5eda-4265-8ade-8eb9b73e51c6.png)
Kind regards,
Arun

**Comment by GiudGiud** [comment]
Hello

You can change the amount of output by setting in the MultiApp input file (sub_gs_uniaxaiaal_noCPFE.i) a different exodus output. 
Then you can set an execute_on to limit the output to certain times in the simulation.
You can also disable output by some of the subapps by passing command line arguments (Outputs/exodus=false) to certain subapps. This is done in the `[MultiApps]` block of the main app. 

Option 1:
Instead of 

```
[Outputs]
  exodus = true
[]
```

you set

```
[Outputs]
  [exo]
    type = Exodus
    execute_on = FINAL
  []
[]
```

Option 2:

```
[MultiApps]
  [sub]
    type = QuadraturePointMultiapp
    input_files = 'sub_gs_uniaxial_noCPFE.i'
    output_in_position = true
    execute_on = timestep_end
    reset_time = '0.02 0.03 0.04 0.05'
    cli_args = 'Outputs/exodus=false' # passed to all of them! otherwise need one param per quadrature point
  []
[]
```

**Reply by abarun22** [reply]
Hi Guillaume,
Option 2 does not create any of the exodus files at all, which is not the preferred case. Option 1 does seem to limit the formation of outputs, except for the first reset_time. This also requires `execute_on = FINAL` to be set at the sub-app output block. One weird thing being that the default outputfiles (e.g.,  main_sub0.e) are not created in this case.

Option: 1
![image](https://user-images.githubusercontent.com/40111479/156429229-309d27df-e0b3-4d05-81ea-d2dce606ee9e.png)

Kind regards,
Arun

**Reply by GiudGiud** [reply]
Option 1:
you can also create a custom execute_on to output only at given times. Will require coding
Option 2:
yes you will need a really long input of exodus=false/true if you want to customize. You can create that with a python script though

**Reply by abarun22** [reply]
Looks like option 1 seems to be the more relevant one. I will see if i can implement something which can only output the default output files suppressing the other ones for times specified in the 'reset_time" list. I may need to priorities other tasks before taking up this work.

**Reply by GiudGiud** [reply]
I think there may be a 3rd option leveraging the Controls system. I ll have to check though and I ll report back

**Reply by GiudGiud** [reply]
so you could use a Control on the Outputs/Exodus block
https://mooseframework.inl.gov/source/controls/ConditionalFunctionEnableControl.html
based on the value of a Function that is the value of a postprocessor
https://mooseframework.inl.gov/source/postprocessors/FunctionValuePostprocessor.html
for that postprocessor you would set the value from main app using this:
https://mooseframework.inl.gov/source/transfers/MultiAppReporterTransfer.html

not super straightforward

**Comment by lynnmunday** [comment]
@abarun22 Have you always been getting the *.e-s002 exodus outputs?  I recently noticed some of these showing up in my simulations that used to only output a single exodus file on the subapp.  Does it output one of these files everytime a reset is hit?

**Comment by abarun22** [comment]
Yes Lynn, i've always been getting these additional exodus outputs whenever
i am resetting my sub-apps through `reset_time` parameter. I recently
enabled this parameter to accept a list of times for which the reset
happens. This triggered a creation of series of files for all the reset
times, that bloated my working directory. Working with Guillaume to find a
solution for this issue, so that only the default 'end of simulation' files
are created.

**Reply by lynnmunday** [reply]
Do you know what reset_app causes these extra output files to be written?  Without reset_app, I don't get seperate exodus output files on my subapp

**Comment by abarun22** [comment, marked as answer]
Yes that's correct. The extra output files will be formed for any app
number that you provide in the 'reset_apps' parameter list. If none is
provided, you will only receive default output files which is good, but at
the cost of app reset. Need to find a balanced way.

**Reply by GiudGiud** [reply]
Ok I think I understand the problem better now. You just want to get rid of the -s002 files. I ll have a look into that. I was just giving general advice into how to reduce the number of exodus files.

**Reply by GiudGiud** [reply]
@abarun22 do you actually need to specify the reset_apps parameter? This parameter is only used to restrict the apps that are being reset. So within a MultiApp with 4 subapps for example, only reset apps 1 and 3 with 'reset_apps = 1 3'
Very often though, all subapps should be reset in which case this parameter need not be specified. 

wrt diagnosing the issue. It seems the reset sets up a 'sequence' output. You can achieve similar output with that parameter of the `Exodus` output.

**Reply by abarun22** [reply]
Hi Guillaume,
Yes, you would need to specify the 'reset_apps' parameter mandatorily to trigger the app reset, without which it will not be possible. The remedy would be to comment out the part of code that creates these extra files whenever user specifies a number under  'reset_apps'. Probbaly it could be an easy fix.
Kind regards,
Arun

**Reply by GiudGiud** [reply]
ah you're right. This is troublesome in itself. You could specify a reset_time and have it be completely ignored if you forgot to specify a reset_apps

**Reply by GiudGiud** [reply]
I fixed that part of the problem in a recent PR (the possibility for a parameter to be ignored)

Once https://github.com/idaholab/moose/pull/20269 or something similar is merged we will have the main problem solved too, you wont need to reset the app, just trigger a recover which is lighter
